### ISSUE TITLE
Make `loadTileMap` accept jsonUrl and image url.

### ISSUE THREAD

**Issue by vgautamm** [notnullgames/tiled-kaboom#2]
Make `loadTileMap` accept jsonUrl and image url.

Currently the plugin breaks.

Here is my proposed solution

```js
const mapsymbols = [...new Array(4052)].reduce((a, v, c) => {
  if (
    [
      92, 173, 847, 888, 889, 907, 909, 930, 1328, 1367, 1368, 1376, 1416, 1419,
      1420, 1421, 1424, 1565, 1867, 1868, 2081, 2082, 2094, 2095, 2111, 2140,
      2141, 2445, 2446, 2449, 2450, 2473, 2481, 2490, 2491, 2501, 2502, 2505,
      2506, 2526, 2532, 2533, 2564, 2577, 2578, 2601, 2609, 2612, 2615, 2618,
      2619, 2621, 2633, 2634, 2638, 2639, 2640, 2653, 2692, 2702, 2706, 2729,
      2737, 2740, 2746, 2747, 2758, 2762, 2766, 2767, 2788, 2789, 2820, 2829,
      2830, 2833, 2834, 2857, 2865, 2868, 2874, 2875, 2885, 2886, 2889, 2890,
      2910, 2916, 2917, 2948, 2955, 2956, 2957, 2961, 2966, 2967, 2968, 2971,
      2973, 2976, 2977, 2978, 2981, 2982, 2983, 2987, 2988, 2989,
    ].includes(c)
  ) {
    return a;
  }

  if (c < 33) {
    return a;
  }

  const ranges = [
    [126, 161],
    [895, 900],
    [1479, 1488],
    [1514, 1520],
    [1524, 1536],
    [1969, 1984],
    [2042, 2048],
    [2142, 2207],
    [2238, 2260],
    [2384, 2389],
    [2482, 2486],
    [2510, 2519],
    [2519, 2524],
    [2555, 2561],
    [2570, 2575],
    [2626, 2631],
    [2641, 2649],
    [2654, 2662],
    [2677, 2689],
    [2768, 2784],
    [2801, 2809],
    [2809, 2817],
    [2893, 2902],
    [2903, 2908],
    [2935, 2946],
  ];

  for (let r of ranges) {
    if (c > r[0] && c < r[1]) {
      return a;
    }
  }

  return [...a, String.fromCharCode(c)];
}, []);

function verifyTiledMap(mapObj) {
  // TODO: add more validation here
  if (!mapObj) {
    throw new Error("Map parameter is required.");
  }

  if (mapObj.orientation !== "orthogonal") {
    throw new Error(
      `Map is ${mapObj.orientation}. Only orthogonal is currently supported.`
    );
  }
  if (mapObj.renderorder !== "right-down") {
    throw new Error(
      `Map is ${mapObj.renderorder}. Only right-down is currently supported.`
    );
  }
}

// decode data in Tiled base64 format
function base64Decode(data) {
  const len = data.length;
  const bytes = new Array(len / 4);

  // Interpret data as an array of bytes representing little-endian encoded uint32 values.
  for (let i = 0; i < len; i += 4) {
    bytes[i / 4] =
      (data.charCodeAt(i) |
        (data.charCodeAt(i + 1) << 8) |
        (data.charCodeAt(i + 2) << 16) |
        (data.charCodeAt(i + 3) << 24)) >>>
      0;
  }

  return bytes;
}

const tiledMapPlugin = (k) => {
  // this will be imported if needed for compressed layers
  let pako = window.pako;

  return {
    verifyTiledMap,

    async loadTiledMap(jsonUrl, imageUrl) {
      return fetch(jsonUrl)
        .then((res) => res.json())
        .then(async (mapObj) => {
          verifyTiledMap(mapObj);

          const map = {
            sprites: [],
            levels: [],
            key: {},
          };

          for (let tileset of mapObj.tilesets) {
            if (tileset.margin !== 0) {
              console.error(
                `${tileset.name} uses margins. That is currently unsupported. Skipping.`
              );
              continue;
            }
            if (tileset.spacing !== 0) {
              console.error(
                `${tileset.name} uses spacing. That is currently unsupported. Skipping.`
              );
              continue;
            }

            const sliceX = tileset.imagewidth / tileset.tilewidth;
            const sliceY = tileset.imageheight / tileset.tileheight;

            map.sprites.push(
              await k.loadSprite(tileset.name, imageUrl, {
                sliceX,
                sliceY,
              })
            );

            for (
              let frame = tileset.firstgid;
              frame < tileset.firstgid + tileset.tilecount;
              frame++
            ) {
              map.key[mapsymbols[frame]] = [
                k.sprite(tileset.name, { frame: frame - 1 }),
              ];
            }
          }

          for (let layer of mapObj.layers) {
            if (layer.type === "tilelayer" && layer.visible) {
              const { width, height } = layer;
              let { data } = layer;

              if (layer.encoding === "base64") {
                if (!layer.compression || layer.compression === "") {
                  data = base64Decode(atob(data));
                } else {
                  if (typeof pako === "undefined") {
                    pako = await import("pako");
                  }
                  const d = Uint8Array.from(
                    atob(data)
                      .split("")
                      .map((c) => c.charCodeAt(0))
                  );

                  console.log("D", d);
                  data = base64Decode(String.fromCharCode(...pako.inflate(d)));
                }
              }

              const mapArray = [...new Array(width)].map(() =>
                new Array(height + 1).fill(" ")
              );
              for (let x = 0; x < width; x++) {
                for (let y = 0; y < height; y++) {
                  const cell = y * width + x;
                  if (data[cell] !== 0) {
                    mapArray[y][x] = mapsymbols[data[cell]];
                  }
                }
              }
              map.levels.push(mapArray.map((l) => l.join("")));
            }
          }

          return map;
        });
    },
  };
};

export default tiledMapPlugin;
```

// usage

const { sprites, levels, key } = await k.loadTiledMap("/map.json", "/img.png")

**Comment by konsumer** [comment]
I am trying to keep the interface simple, and usable in multiple js environments. This solution relies on ` fetch(jsonUrl)` which would not work in all environments. Since the matter of "turn this JSON file into a JS map-object" is super-simple, in just about any JS environment, I think it's better to just get the JSON object, however works best for you, as a user.

I can recommend a few different ways to get a map object, all of which have their own strengths/weaknesses, depending on where it is being used.

## fetch

This is sort of what yours uses, but this is simpler:

```js
const map = await fetch('/maps/mymap.json').then(r => r.json())
```

This will work only directly in browser, and for browserify/webpack/etc it will require you to put files in a specific place that is different than the code. That might be ok for some, but not everyone. The lazy-load aspect is a bonus in my opinion, though, as you could just load a map when you need it.

## inline

Since it's a regular JS-object, you can also just put the map inline, in your code, if you want. I personally would not do this, with all the other, more readable options, but I like that it's a choice a user could make if they want to.

## builder

If you are using webpack or browserify, you can import the file directly:

```js
import map from './maps/map1.json'
```

or

```js
const map = require('./maps/map1.json')
```

depending on if your builder supports commonjs or es6 imports.

It will inline the JSON object when it builds the site, which may be a good or bad thing (lazy-load is cool, but it incurs a delay, this has no delay, but also all maps are loaded up-front, even if they are never used, so it uses more ram, etc.)

## import/assert

Brand new browsers have native ES6 import, and a new keyword `assert` to tell the browser how to load non-js-modules. I use this in [the demo](https://notnullgames.github.io/tiled-kaboom/).

```js
import map from './map.json' assert { type: "json" }
```

Since this can use browser-supported `importmap`, too, you could keep your config in a central place, and just change the importmap, and not the source-code, to move things around. For a no-build setup that targets modern browsers, this is my personal favorite. You can co-locate logic for loading the map, objects in the top-level file that sets up the map, and only import when you need it (lazy-load) so in my opinion, you get the best of all the worlds, with no build required.

A downside: this only works in very up-to-date browsers, but older browsers can also be polyfilled, if you want this style.


> **Sidenote:** This is not a great way to submit a PR. It's much better to fork a project, make your changes, then submit a proper Pull Request, so the maintainer can see the diff that your code changes, very quickly, and see a history of how it was modified, as well as have a discussion about the code, and iterate on change-requests.

**Comment by konsumer** [comment]
It does bring up another concern I had, though, while I was making it. I use `location` param in `loadTiledMap()` to be able to resolve the location of images and eventually, I plan to support non-inline tilesets, and this will allow the user to dynamically load them from a URL, like it does now, with images.

It occurs to me that this all won't really work for every use-case maps currently support, anyway. 

So, now, you set the `location` to a URL prefix (even relative, or remote) and that tells it how to find the images referenced in the map, but if you're doing that anyway, maybe it makes sense to make it all-fetch, too, then later when I add support for external tilemaps, they will all just assume you can get to the referenced files via the `location` prefix.

A simple solution that would resolve both sorts of use-cases (you already have a map object, or you want to fetch it), would be to check the type of map. If it's `string`, do a fetch, if it's an object, use it as-is.

**Comment by konsumer** [comment]
In the interest of supporting both use-cases, I just added it. Now you can use an object or a URL. Published in tiled-kaboom@0.0.10 (on npm) and should update in all other places, shortly. Feel free to comment more, if there is more to talk about, but I am going to close for now.
